platform-sdk/docs/proposals/queuestate-fcqueue-to-virtualmap/proposal.md

```diff
@@ -0,0 +1,159 @@
+
+No changes to `ReadableQueueState` and `WritableQueueState` interfaces.
+
+No changes to `ReadableQueueStateImpl` and `WritableQueueStateImpl` classes.
+
+`QueueNode`:
+
+* No longer marked as `@Deprecated`
+* Has two children: `QueueNodeState` (was: `StringLeaf`) and `VirtualMap` (was: `FCQueue`)
+
+`QueueNodeState`:
+
+* A simple leaf node similar to `StringLeaf`, but contains two long values in addition to
+  a string value
+* These two longs are the head and the tail of the queue - see below
+
+### Queue operations
+
+The underlying `VirtualMap` will map longs to E, where E is the type of the queue. The longs
+will be used to organize the links between queue elements: if an element has position N, then
+its next element has position N+1. When an empty queue is created, its head and tail are both
+pointing to 0. The head is _the next_ position to add, the tails is _the current_ position to
+remove. If head and tail positions are equal, the queue is empty.
+
+![Map keys](map-keys.svg)
+
+**Add**:
+
+* Insert an element to the head position
+* Increment the head by one
+* Code snippet:
+
+```java
+public void add(final E element) {
+    map.put(head++, element);
+}
+```
+
+![Add to the queue](queue-add.svg)
+
+**Poll / remove**:
+
+* Check if the queue is empty
+* Remove the element from the tail position
+* Increment the tail by one
+* Code snippet:
+
+```java
+public E remove() {
+    if (tail == head) {
+        return null;
+    }
+    return map.remove(tail++);
+}
+```
+
+![Remove from the queue](queue-remove.svg)
+
+### Serialization, deserialization, hashing, reconnects
+
+All provided by `VirtualMap`.
+
+`QueueNodeState` will write both its long values in addition to the label during serialization.
+
+### Data serialization and deserialization
+
+`FCQueue` today stores all data items in memory. When the queue is written to a snapshot, all
+items are serialized as `SelfSerializable`. To store items in `VirtualMap`, key and value
+serializers are to be provided. The key serializer will be trivial, as keys are longs. The value
+serializer will be implemented using the codec provided by `QueueNode`, similar to how key/value
+states work.
+
+## Data migration
+
+Current production states contain a few queue state nodes, all are tiny except transaction
+receipts. However, even the largest queue is expected to be small enough, so it can be migrated
+to a virtual map during Hedera node software update window.
+
+Migration code will be in `QueueNode.migrate()` This method will get its right child (which is
+`FCQueue`), iterate over it, insert all elements in the right order to a virtual map, then set
+the map as the new right child. The left child will be changed from a simple string leaf to a
+`QueueNodeState` leaf node.
+
+## Testing
+
+This feature doesn't add any new functionality. All existing Hedera unit tests should pass with
+the new queue state implementation. Unit tests specific to `QueueNode` may need to be updated.
```

Comment:
`FCQueue` allows for concurrent updates (`add`/`remove`). I believe that is tested. Either that requirement should be dropped (as with `VirtualMap`) or a layer of synchronized methods on top of `VirtualMap` should be added. The latter will have performance impact.

Reply:
The way how `FCQeueue` is thread safe is all its methods like add() or remove() are synchronized. As the simplest solution, similar synchronization can be added on top of `VirtualMap`, with similar performance impact.

However, first of all it would be great to understand how queue states are used today. Are they really updated from multiple threads in parallel? @tinker-michaelj @Neeharika-Sompalli your input would be much appreciated here

Reply:
They couldn't possibly be updated from multiple threads, as that would lead to non-determinism in the state. It must be that it is read from multiple threads but only updated by one.

Reply:
Removing the synchronization from the existing implementation sounds risky to me.

Also, I'm not at all clear as to how the current queues achieve determinism. Synchronized `add`/`remove` methods alone won't do the trick because parallel threads can obtain the lock in random order. I'd suggest to study the current usages of the `add`/`remove` methods to understand this aspect (and/or to bug Services people about it), to ensure the new implementation doesn't break existing logic w.r.t. parallel execution.

Reply:
No, `synchronized` methods do not provide determinism. It is achieved by the mode of use (all updates are done by a single thread, namely, transaction handling thread), not by implementation.
The `FCQueue` class, unlike `VirtualMap`, was initially designed as a general purpose data structure to be used in multithreaded environment (and that was tested by unit tests). This requirement, however, is not needed for Services code where `FCQueues` are part of the state. Is there any other use? If yes (which I doubt), that should be left alone and not converted to `VirtualMap`.

Reply:
Just to clarify: are you saying that the transaction handling thread handles all transactions in the consensus (or any other deterministic) order? This seems to be the only way to achieve determinism for the queues.

If so, then yes, I agree that we don't want to implement synchronized methods on the virtual map since they won't really be useful.

Reply:
That's right, all state changes are performed by the transaction handling thread when processing transactions in the consensus order.